velox/functions/lib/Re2Functions.cpp

```diff
@@ -404,26 +404,109 @@ bool matchSuffixPattern(
           length) == 0;
 }
 
+// Search for 'pattern' in 'input' starting from positions greater than
+// 'startPosition'.
+vector_size_t matchPatternAtPosition(
+    const StringView& input,
+    const StringView& pattern,
+    vector_size_t startPosition) {
+  const auto inputData = input.data();
+  const auto patternData = pattern.data();
+  const auto length = pattern.size();
+  auto searchStartIdx = startPosition;
+  if (input.size() - startPosition < length) {
+    return -1;
+  }
+  const auto lastPosition = input.size() - length;
+
+  while (searchStartIdx <= lastPosition) {
+    if (std::memcmp(inputData + searchStartIdx, patternData, length) == 0) {
+      return searchStartIdx;
+    }
+    searchStartIdx++;
+  }
+  return -1;
+}
+
 template <PatternKind P>
 class OptimizedLikeWithMemcmp final : public VectorFunction {
  public:
   OptimizedLikeWithMemcmp(
-      StringView pattern,
-      vector_size_t reducedPatternLength)
-      : pattern_{pattern}, reducedPatternLength_{reducedPatternLength} {}
+      const StringView& pattern,
+      const PatternMetadata& patternParameters)
+      : pattern_{pattern}, patternMetadata_{patternParameters} {
+    trailingWildcard_ = pattern_.data()[pattern_.size() - 1] == '%';
+  }
+
+  // Match 'numFixedPatterns' fixed patterns from the vector of fixedPatterns
+  // in the pattern with input in their order of occurence.
+  vector_size_t matchNFixedPatterns(
+      const StringView& input,
+      const vector_size_t numFixedPatterns) const {
+    const auto fixedPatterns = patternMetadata_.fixedPatterns.value();
+    auto startPosition = 0;
+
+    for (auto i = 0; i < numFixedPatterns; i++) {
+      const std::string fixedPatternString{fixedPatterns[i]};
+      auto fixedPattern = StringView(fixedPatternString);
+      if ((startPosition = matchPatternAtPosition(
+               input, fixedPattern, startPosition)) == -1) {
+        return -1;
+      }
+      startPosition += fixedPattern.size();
+    }
+    return startPosition;
+  }
+
+  // Match all fixed patterns in an input string having fixed patterns
+  // interspersed with one or more '%' wildcard character streams.
+  bool matchFixedWithWildcard(const StringView& input) const {
+    auto startPosition = 0;
+    const auto numFixedPatterns = patternMetadata_.numFixedPatterns;
+
+    if (!trailingWildcard_) {
```

Comment:
The way I understand this algorithm is that when matching there is forward matching from left to right and backward matching from right to left. If !trailingWildcard_ then the last pattern match has to be from right to left.

Its easier to write this code as 

int numMatches = trailingWildcard_ ? numFixedPatterns : numFixedPatterns - 1;
call MatchFixedPatterns
 
if (trailingWildcard_) {
.. code for last backward match...
}

Reply:
Thanks for the suggestion!